### PR TITLE
Added a condition to find the latest compatible version

### LINKED ISSUE #15745: Incorrect parsing of latest version.json shows old version as latest
;)

### PR DESCRIPTION
**Pull request by Gaurav-Punjabi**
Signed-off-by: Gaurav Punjabi <gaurav.bpunjabi@gmail.com>

### Description
Added a condition getLatestCompatibleVersion() to fetch the latest compatible version.

Please describe your pull request.

Fixes #15745 

Before submitting pull request, please review the following checklist:

- [x] Make sure you have read our [CONTRIBUTING.md](https://github.com/phpmyadmin/phpmyadmin/blob/master/CONTRIBUTING.md) document.
- [x] Make sure you are making a pull request against the correct branch. For example, for bug fixes in a released version use the corresponding QA branch and for new features use the _master_ branch. If you have a doubt, you can ask as a comment in the bug report or on the mailing list.
- [x] Every commit has proper `Signed-off-by` line as described in our [DCO](https://github.com/phpmyadmin/phpmyadmin/blob/master/DCO). This ensures that the work you're submitting is your own creation.
- [x] Every commit has a descriptive commit message.
- [x] Every commit is needed on its own, if you have just minor fixes to previous commits, you can squash them.
- [x] Any new functionality is covered by tests.

### CODE REVIEW

**Comment by ibennetch** [comment]
I seem to get this error in my logs after applying your patch. My current branch is QA_5_0, so my current version is 5.0.1-dev and the current downloads offered in version.json are 4.9.3 and 5.0.0.

> PHP Fatal error:  Uncaught Error: Cannot use object of type stdClass as array in /var/www/pma-dev/version_check.php:40\nStack trace:\n#0 {main}\n  thrown in /var/www/pma-dev/version_check.php on line 40

**Comment by Gaurav-Punjabi** [comment]
> I seem to get this error in my logs after applying your patch. My current branch is QA_5_0, so my current version is 5.0.1-dev and the current downloads offered in version.json are 4.9.3 and 5.0.0.
> 
> > PHP Fatal error:  Uncaught Error: Cannot use object of type stdClass as array in /var/www/pma-dev/version_check.php:40\nStack trace:\n#0 {main}\n  thrown in /var/www/pma-dev/version_check.php on line 40

Fixed it 👍

File: test/classes/VersionInformationTest.php [hunk 1]
```diff
@@ -288,6 +295,48 @@ public function testGetLaestCompatibleVersionWithOldPHPVersion()
         $this->assertEquals('4.0.10.10', $compatible['version']);
     }
 
+
+    /**
+     * Tests getLatestCompatibleVersion() with an new PHP version
+     *
+     * @return void
+     */
+    public function testGetLaestCompatibleVersionWithNewPHPVersion()
```
Comment (mauriciofauth) [inline test/classes/VersionInformationTest.php]:
```suggestion
    public function testGetLatestCompatibleVersionWithNewPHPVersion()
```

**Review by MauricioFauth — CHANGES_REQUESTED** [review]
Thanks for your contribution.
Tests are failing. See: https://travis-ci.org/phpmyadmin/phpmyadmin/jobs/633798240#L430

File: test/classes/VersionInformationTest.php (hunk 1 above)
Comment (Gaurav-Punjabi) [reply test/classes/VersionInformationTest.php]: Wait, I'll fix these.

**Comment by ecartz** [comment]
```
            // To compare the current release with the previous latest release
            if ($latestRelease) {
                if ($latestRelease['version'] < $release->version) {
                    $latestRelease = [
                        'version' => $release->version,
                        'date' => $release->date
                    ];
                }
            } else {
                $latestRelease = [
                    'version' => $release->version,
                    'date' => $release->date,
                ];
            }
        }
        // no compatible version
        return $latestRelease;
```
can more easily be written 
```
            if (is_null($latestRelease) || $latestRelease['version'] < $release->version) {
                $latestRelease = [
                  'version' => $release->version,
                  'date' => $release->date,
                ];
            }
        }

        return $latestRelease;
```
Then it wouldn't have the code duplicated.  And it wouldn't say "no compatible version" right before it returns the compatible version in the most common case.  I tested this version on my 5.0.1 installation.  

Also, wouldn't this be more readable with the vertical spaces before the comments as used to be there?  

It's not clear to me if this is the correct solution.  If I'm currently on 4.9.3, what should it say?  Upgrade to 5.0.1?  Or to 4.9.4?  Or both?  If I'm on 4.9.4, should it be pushing me to 5.0.1 or not?  I.e. this fixes the error of showing 4.9.4 as the latest version when on 5.0.1 (which was the problem that I was experiencing), but it may introduce other errors.  

I think that if I'm currently on 

* 5.0.1-Snapshot, it should say that I have a dev version with 5.0.1 the latest stable.  And maybe it should tell me if I have the latest dev version or not (is there a newer snapshot?).  
* 5.0.1, it should say I have the latest version.  
* 5.0.0, it should say that there is a newer version, 5.0.1.  
* 4.9.4, it should say that is LTS with 5.0.1 stable.  
* 4.9.3, it should say that there is 4.9.4 LTS and 5.0.1 stable.  
* 4.8.* or older, it should say that it is out of date and needs updated.  Not sure what version to recommend.  

I think that latest release is itself a weak concept that doesn't quite match actual usage.  Security-wise, 4.9.4 is probably a bit safer than 5.0.1, as it gets all the security fixes and none of the new features that might require security fixes.  So in my opinion, it shouldn't say that 4.9.4 needs updated to 5.0.1, just that that option is available. But the way it works now, it will always suggest converting to one version for everyone.  With the buggy behavior that it was telling people to downgrade from 5.0.1 to 4.9.4.

**Comment by ibennetch** [comment]
> Then it wouldn't have the code duplicated.

That's great.

> 5.0.1-Snapshot, it should say that I have a dev version with 5.0.1 the latest stable. And maybe it should tell me if I have the latest dev version or not (is there a newer snapshot?).

`5.0.1-dev, latest release is 5.0.1` because the user is running a version more recent than the latest stable release.

> 5.0.1, it should say I have the latest version.

Correct

> 5.0.0, it should say that there is a newer version, 5.0.1.

I agree

> 4.9.4, it should say that is LTS with 5.0.1 stable.

I don't know if we currently provide enough information in version.json to display this. If we were to add this, `Latest LTS; current stable: 5.0.1` may be a good way to handle it.

> 4.9.3, it should say that there is 4.9.4 LTS and 5.0.1 stable.

Hm, this one is a bit more difficult. We don't really want people to stay on 4.9, but if someone decides to stay on 4.9 then perhaps it would be good to announce if a new LTS version is available. I don't think we've encountered this specific situation before since the version check feature has been introduced.

I slightly disagree; we're recommending people migrate to 5.x because 4.9 will eventually be deprecated, and for best compatibility with recent MySQL/MariaDB and PHP versions users should upgrade to 5.0, but I can see the case for telling people that a new LTS version is available. I don't know if we currently provide enough metadata in version.json to make that determination.

We historically haven't done much to tell people about LTS version support; we have several other notification methods users can subscribe through to find these announcements. This has been used to show the latest _stable_ version, and I'm not sure that fixing that is going to be easy.

> 4.8.* or older, it should say that it is out of date and needs updated. Not sure what version to recommend.

**Comment by Gaurav-Punjabi** [comment]
> ```
>             // To compare the current release with the previous latest release
>             if ($latestRelease) {
>                 if ($latestRelease['version'] < $release->version) {
>                     $latestRelease = [
>                         'version' => $release->version,
>                         'date' => $release->date
>                     ];
>                 }
>             } else {
>                 $latestRelease = [
>                     'version' => $release->version,
>                     'date' => $release->date,
>                 ];
>             }
>         }
>         // no compatible version
>         return $latestRelease;
> ```
> 
> can more easily be written
> 
> ```
>             if (is_null($latestRelease) || $latestRelease['version'] < $release->version) {
>                 $latestRelease = [
>                   'version' => $release->version,
>                   'date' => $release->date,
>                 ];
>             }
>         }
> 
>         return $latestRelease;
> ```
> 
> Then it wouldn't have the code duplicated. And it wouldn't say "no compatible version" right before it returns the compatible version in the most common case. I tested this version on my 5.0.1 installation.
> 
> Also, wouldn't this be more readable with the vertical spaces before the comments as used to be there?
> 
> It's not clear to me if this is the correct solution. If I'm currently on 4.9.3, what should it say? Upgrade to 5.0.1? Or to 4.9.4? Or both? If I'm on 4.9.4, should it be pushing me to 5.0.1 or not? I.e. this fixes the error of showing 4.9.4 as the latest version when on 5.0.1 (which was the problem that I was experiencing), but it may introduce other errors.
> 
> I think that if I'm currently on
> 
> * 5.0.1-Snapshot, it should say that I have a dev version with 5.0.1 the latest stable.  And maybe it should tell me if I have the latest dev version or not (is there a newer snapshot?).
> * 5.0.1, it should say I have the latest version.
> * 5.0.0, it should say that there is a newer version, 5.0.1.
> * 4.9.4, it should say that is LTS with 5.0.1 stable.
> * 4.9.3, it should say that there is 4.9.4 LTS and 5.0.1 stable.
> * 4.8.* or older, it should say that it is out of date and needs updated.  Not sure what version to recommend.
> 
> I think that latest release is itself a weak concept that doesn't quite match actual usage. Security-wise, 4.9.4 is probably a bit safer than 5.0.1, as it gets all the security fixes and none of the new features that might require security fixes. So in my opinion, it shouldn't say that 4.9.4 needs updated to 5.0.1, just that that option is available. But the way it works now, it will always suggest converting to one version for everyone. With the buggy behavior that it was telling people to downgrade from 5.0.1 to 4.9.4.
Okay i'll make these changes in the code to make it clean.

**Comment by williamdes** [comment]
For the LTS/security part see #12506

**Comment by williamdes** [comment]
And yes I did commit this to the main repo in d26b9fb543cfe63c7383585e9b9785fcf1cf6920 and then revert by 87aa7b0dc7e0913f4f4eed8f30475a8c37d9e050

Sorry.

**Comment by williamdes** [comment]
Hi everyone
I intend to merge this PR today, could you please review it ?
NB: you can use the approve button :)

**Review by boherm — APPROVED** [review]
It looks good to me :)

**Comment by williamdes** [comment]
Thank you @Gaurav-Punjabi for the initial work on this issue !

**Comment by Gaurav-Punjabi** [comment]
> Thank you @Gaurav-Punjabi for the initial work on this issue !

Glad I could help.

**Pull request merged**